File: tools/layers.py
"""Split a drawing into layers before asking anything else of it.

Every hard bug this tracer has had came from one cause: it looked at the sheet
as a single undifferentiated pile of geometry and tried to tell a wall from a
counter by shape. The drawing already separates them and says so in the pen.

    the window FRAME paired like a wall and plugged all three windows
    the sink COUNTER paired with the half-wall line and made a wall
    a 3in mark inside the wall counted as a jamb and ate a picture window

None of those is a geometry problem. A frame, a counter and a vent are drawn
with different pens from the wall they sit in, and every one of those failures
would have been impossible if the wall layer had been isolated first.

## The key

A layer is (stroke colour, stroke width, fill colour). All three, because no
one of them separates on its own -- proven on two drafters who agree on nothing:

    Southern Integrity   every pen 0.5pt wide, separated ENTIRELY by colour
                         #000000 walls, #804040 plumbing, #4b4b4b cabinets,
                         #6d86a9 elevation markers, #004080 dimensions
    SketchUp LayOut      every pen black, separated ENTIRELY by width
                         0.2 / 0.4 / 0.6 / 0.8 / 1.0 / 1.2 / 1.8

Key on colour alone and LayOut collapses to one layer. Key on width alone and
Southern Integrity does. Key on both and each drafter's convention falls out
without either being named.

Fill belongs in the key too, and is not decoration: Jay Osborne poche-fills his
walls solid #646464, while Southern Integrity leaves walls hollow and fills the
CABINETS. The same "is it filled" question gives opposite answers on the two
sheets, so it has to be observed per drawing rather than assumed.

## What this does and does not do

It splits, measures and describes. It does NOT decide which layer is the wall
layer -- see rank(), which is a first attempt and is meant to be checked
against layer_check.py's renders rather than trusted. Getting the split right
is worth having on its own: even hand-picking a layer by eye from a labelled
list beats the tracer guessing from shape, which is what it does today.
"""

import math
from collections import defaultdict

# Segments shorter than this are noise for measurement purposes -- hatching
# ends, glyph fragments. They stay in the layer, they just do not vote.
MIN_MEASURE_IN = 1.0
# Off-axis by less than this and a segment counts as axis-aligned.
AXIS_TOL = 0.3
# Parallel pair gaps outside this range tell us nothing about wall thickness.
PAIR_MIN_IN = 2.0
PAIR_MAX_IN = 14.0
# Two faces must run alongside each other this far to count as a pair.
PAIR_OVERLAP_IN = 18.0


def _overlaps(clip, rect):
    """Do these rectangles share any ground?

    Not Rect.intersects(): pymupdf calls a zero-AREA rectangle empty and
    reports no intersection for it, and a path holding a single horizontal or
    vertical line has exactly that -- zero height or zero width. So the test
    silently threw away every axis-aligned line on the sheet. On the kitchen
    plan it cut 10600 black-stroked paths down to 36, which is the wall layer
    reduced to nothing while every hatched and filled thing survived intact.
    """
    return (rect.x1 >= clip.x0 and rect.x0 <= clip.x1
            and rect.y1 >= clip.y0 and rect.y0 <= clip.y1)


def _hex(colour):
    if colour is None:
        return None
    return "#%02x%02x%02x" % tuple(int(round(c * 255)) for c in colour)


def split(page, clip, scale):
    """Every path in `clip`, grouped by (stroke colour, width, fill colour).

    Geometry comes back in real inches, already sorted into horizontal runs,
    vertical runs, diagonals and curves -- which is what every consumer wants
    and what each of them currently re-derives for itself.
    """
    layers = {}
    for path in page.get_drawings():
        if not _overlaps(clip, path["rect"]):
            continue
        key = (_hex(path.get("color")),
               round(path.get("width") or 0.0, 2),
               _hex(path.get("fill")))
        layer = layers.get(key)
        if layer is None:
            layer = layers[key] = {
                "colour": key[0], "width": key[1], "fill": key[2],
                "horizontal": [], "vertical": [], "diagonal": [],
                "curves": 0, "paths": 0, "rects": [],
            }
        layer["paths"] += 1
        if path.get("fill") is not None:
            rect = path["rect"]
            layer["rects"].append((rect.x0 * scale, rect.y0 * scale,
                                   rect.x1 * scale, rect.y1 * scale))
        for item in path["items"]:
            if item[0] == "c":
                layer["curves"] += 1
                continue
            if item[0] != "l":
                continue
            a, b = item[1], item[2]
            if not (clip.contains(a) and clip.contains(b)):
                continue
            ax, ay, bx, by = (a.x * scale, a.y * scale, b.x * scale, b.y * scale)
            if abs(ay - by) < AXIS_TOL and abs(ax - bx) >= AXIS_TOL:
                layer["horizontal"].append((ay, min(ax, bx), max(ax, bx)))
            elif abs(ax - bx) < AXIS_TOL and abs(ay - by) >= AXIS_TOL:
                layer["vertical"].append((ax, min(ay, by), max(ay, by)))
            elif abs(ax - bx) >= AXIS_TOL or abs(ay - by) >= AXIS_TOL:
                layer["diagonal"].append((ax, ay, bx, by))
    for layer in layers.values():
        describe(layer)
    return layers


def describe(layer):
    """Measure a layer, so it can be judged without being looked at."""
    runs = layer["horizontal"] + layer["vertical"]
    axis_ink = sum(hi - lo for _, lo, hi in runs if hi - lo >= MIN_MEASURE_IN)
    diagonal_ink = sum(math.hypot(bx - ax, by - ay)
                       for ax, ay, bx, by in layer["diagonal"]
                       if math.hypot(bx - ax, by - ay) >= MIN_MEASURE_IN)

    xs = [c for c, _, _ in layer["vertical"]] + \
         [v for _, lo, hi in layer["horizontal"] for v in (lo, hi)]
    ys = [c for c, _, _ in layer["horizontal"]] + \
         [v for _, lo, hi in layer["vertical"] for v in (lo, hi)]

    layer["segments"] = len(runs) + len(layer["diagonal"])
    layer["ink_in"] = axis_ink + diagonal_ink
    layer["axis_share"] = axis_ink / (axis_ink + diagonal_ink) if axis_ink + diagonal_ink else 0.0
    layer["extent"] = ((min(xs), min(ys), max(xs), max(ys))
                       if xs and ys else None)
    layer["pairs"] = pair_gaps(layer)
    return layer


def pair_gaps(layer):
    """Gaps between parallel runs that lie alongside each other, tallied.

    A wall layer piles these up on one or two values -- the thicknesses that
    drafter builds in -- because a wall is two faces a construction thickness
    apart, held over the whole sheet. A cabinet layer scatters them across
    every depth in the kitchen, and an annotation layer produces none at all.
    That difference is what makes a layer's role measurable rather than named.
    """
    tally = defaultdict(float)
    for runs in (layer["horizontal"], layer["vertical"]):
        ordered = sorted(runs)
        for index, (c1, lo1, hi1) in enumerate(ordered):
            for c2, lo2, hi2 in ordered[index + 1:]:
                gap = c2 - c1
                if gap > PAIR_MAX_IN:
                    break
                if gap < PAIR_MIN_IN:
                    continue
                overlap = min(hi1, hi2) - max(lo1, lo2)
                if overlap >= PAIR_OVERLAP_IN:
                    tally[round(gap * 2) / 2] += overlap
    return dict(tally)


def concentration(layer):
    """How much of a layer's pairing sits on its single commonest gap.

    High means one repeated thickness, which is construction. Low means a
    scatter of depths, which is cabinetry and fixtures.
    """
    tally = layer["pairs"]
    if not tally:
        return 0.0
    return max(tally.values()) / sum(tally.values())


def rank(layers, page_area_sqin):
    """A first guess at which layer is which. Check it, do not trust it.

    Deliberately a separate step from split(), and deliberately weak. The
    evidence for these rules is two drafters, which is exactly the sample size
    that produced "colour first, weight second" earlier in this project -- a
    rule the LayOut sheets then demolished. Treat the score as a shortlist for
    a human, not a decision.
    """
    scored = []
    for key, layer in layers.items():
        if layer["segments"] < 20 or not layer["extent"]:
            continue
        x0, y0, x1, y1 = layer["extent"]
        spread = ((x1 - x0) * (y1 - y0)) / page_area_sqin if page_area_sqin else 0.0
        conc = concentration(layer)
        # Structure: axis-aligned, spread across the plan, and pairing up on a
        # repeated thickness. All three, because any two of them also describe
        # a run of kitchen cabinets.
        structure = (layer["axis_share"] * min(spread * 3.0, 1.0)
                     * conc * math.log1p(layer["ink_in"]))
        scored.append({
            "key": key, "layer": layer, "spread": spread,
            "concentration": conc, "structure": structure,
            "modal_gap": (max(layer["pairs"].items(), key=lambda kv: kv[1])[0]
                          if layer["pairs"] else None),
        })
    return sorted(scored, key=lambda s: -s["structure"])

File: app/src/app/import/layers.js
// @ts-check
/**
 * Split a sheet by pen, and rank which pen drew the structure.
 *
 * A port of `tools/layers.py`. The rules and the failures behind them are
 * recorded in AGENTS.md under "Split by pen before you measure anything" and
 * are not repeated here; what follows is only what a reader of THIS file needs.
 *
 * A layer is **(stroke colour, stroke width, fill)** - all three. Neither
 * colour nor width alone separates, proven on two drafters who share no
 * conventions: one sheet is every pen at 0.5pt separated entirely by colour,
 * the other every pen black separated entirely by width. Fill belongs in the
 * key and is not decoration - one drafter poche-fills walls, another fills
 * cabinets and leaves walls hollow.
 *
 * Geometry comes out in REAL INCHES, already sorted into horizontal runs,
 * vertical runs and diagonals, because that is what every consumer wants.
 */

import {containsPoint, hexOf, overlaps} from './readers/model.js';
import {pyRound} from './round.js';

/** A run shorter than this is not measured; it is a tick or a corner. */
const MIN_MEASURE_IN = 1.0;
/** How far off axis a segment may be and still count as axis aligned. */
const AXIS_TOL = 0.3;

/** The gap band `pair_gaps` tallies over. Wide, since it is only evidence. */
const PAIR_MIN_IN = 2.0;
const PAIR_MAX_IN = 14.0;
/** Two runs must lie alongside each other this far to be a pair. */
const PAIR_OVERLAP_IN = 18.0;

/** The band a layer's commonest gap must fall in to look like construction. */
const WALL_GAP_MIN_IN = 2.5;
const WALL_GAP_MAX_IN = 8.5;

/** A layer with fewer segments than this is not judged at all. */
const MIN_SEGMENTS = 20;

/**
 * @typedef {object} Layer
 * @property {?string} colour
 * @property {number} width
 * @property {?string} fill
 * @property {Array<number[]>} horizontal `[coord, lo, hi]`, in inches.
 * @property {Array<number[]>} vertical
 * @property {Array<number[]>} diagonal `[ax, ay, bx, by]`.
 * @property {number} curves
 * @property {number} paths
 * @property {Array<number[]>} rects
 * @property {number} segments
 * @property {number} ink
 * @property {number} axisShare
 * @property {?number[]} extent
 * @property {Record<string, number>} pairs
 */

/**
 * The key a layer is filed under. A string because it is a map key, and
 * because it is exactly what `--layer "#000000,0.5"` types on the command
 * line.
 *
 * @param {?string} colour
 * @param {number} width
 * @param {?string} fill
 */
export function layerKey(colour, width, fill)
{
	return `${colour || 'none'},${width},${fill || 'none'}`;
}

/**
 * Parse a layer key, as `--layer` does: `"#000000,0.5"` or `"none,0,#646464"`.
 *
 * @param {string} text
 */
export function parseKey(text)
{
	const parts = text.split(',').map((part) => part.trim());
	while (parts.length < 3)
	{
		parts.push('none');
	}
	const blank = (part) => ['none', '-', ''].includes(part.toLowerCase());
	return layerKey(
		blank(parts[0]) ? null : parts[0],
		blank(parts[1]) ? 0 : Number(parts[1]),
		blank(parts[2]) ? null : parts[2]);
}

/**
 * Every path in `clip`, grouped by pen.
 *
 * @param {import('./readers/model.js').DrawnPath[]} paths
 * @param {number[]} clip In points.
 * @param {number} scale Real inches per point.
 * @returns {Map<string, Layer>}
 */
export function split(paths, clip, scale)
{
	/** @type {Map<string, Layer>} */
	const layers = new Map();

	for (const path of paths)
	{
		if (!overlaps(clip, path.rect))
		{
			continue;
		}
		const colour = hexOf(path.colour);
		const width = pyRound(path.width || 0, 2);
		const fill = hexOf(path.fill);
		const key = layerKey(colour, width, fill);

		let layer = layers.get(key);
		if (!layer)
		{
			layer = {
				colour, width, fill,
				horizontal: [], vertical: [], diagonal: [],
				curves: 0, paths: 0, rects: [],
				segments: 0, ink: 0, axisShare: 0, extent: null, pairs: {},
			};
			layers.set(key, layer);
		}
		layer.paths += 1;
		if (path.fill)
		{
			layer.rects.push([
				path.rect[0] * scale, path.rect[1] * scale,
				path.rect[2] * scale, path.rect[3] * scale,
			]);
		}

		for (const item of path.items)
		{
			if (item[0] === 'c')
			{
				layer.curves += 1;
				continue;
			}
			if (item[0] !== 'l')
			{
				continue;
			}
			const a = item[1];
			const b = item[2];
			if (!containsPoint(clip, a) || !containsPoint(clip, b))
			{
				continue;
			}
			const ax = a.x * scale;
			const ay = a.y * scale;
			const bx = b.x * scale;
			const by = b.y * scale;
			if (Math.abs(ay - by) < AXIS_TOL && Math.abs(ax - bx) >= AXIS_TOL)
			{
				layer.horizontal.push([ay, Math.min(ax, bx), Math.max(ax, bx)]);
			}
			else if (Math.abs(ax - bx) < AXIS_TOL && Math.abs(ay - by) >= AXIS_TOL)
			{
				layer.vertical.push([ax, Math.min(ay, by), Math.max(ay, by)]);
			}
			else if (Math.abs(ax - bx) >= AXIS_TOL || Math.abs(ay - by) >= AXIS_TOL)
			{
				layer.diagonal.push([ax, ay, bx, by]);
			}
		}
	}

	for (const layer of layers.values())
	{
		describe(layer);
	}
	return layers;
}

/**
 * Measure a layer, so it can be judged without being looked at.
 *
 * @param {Layer} layer
 */
export function describe(layer)
{
	const runs = layer.horizontal.concat(layer.vertical);
	let axisInk = 0;
	for (const [, lo, hi] of runs)
	{
		if (hi - lo >= MIN_MEASURE_IN)
		{
			axisInk += hi - lo;
		}
	}
	let diagonalInk = 0;
	for (const [ax, ay, bx, by] of layer.diagonal)
	{
		const length = Math.hypot(bx - ax, by - ay);
		if (length >= MIN_MEASURE_IN)
		{
			diagonalInk += length;
		}
	}

	const xs = layer.vertical.map(([c]) => c)
		.concat(layer.horizontal.flatMap(([, lo, hi]) => [lo, hi]));
	const ys = layer.horizontal.map(([c]) => c)
		.concat(layer.vertical.flatMap(([, lo, hi]) => [lo, hi]));

	layer.segments = runs.length + layer.diagonal.length;
	layer.ink = axisInk + diagonalInk;
	layer.axisShare = (axisInk + diagonalInk) ? axisInk / (axisInk + diagonalInk) : 0;
	layer.extent = (xs.length && ys.length)
		? [Math.min(...xs), Math.min(...ys), Math.max(...xs), Math.max(...ys)]
		: null;
	layer.pairs = pairGaps(layer);
	return layer;
}

/**
 * Gaps between parallel runs that lie alongside each other, tallied.
 *
 * A wall layer piles these onto one or two values - the thicknesses that
 * drafter builds in - because a wall is two faces a construction thickness
 * apart, held over the whole sheet. A cabinet layer scatters them across every
 * depth in the kitchen; an annotation layer produces none.
 *
 * @param {Layer} layer
 * @returns {Record<string, number>}
 */
export function pairGaps(layer)
{
	/** @type {Record<string, number>} */
	const tally = {};
	for (const runs of [layer.horizontal, layer.vertical])
	{
		const ordered = sortRuns(runs);
		for (let index = 0; index < ordered.length; index += 1)
		{
			const [c1, lo1, hi1] = ordered[index];
			for (let other = index + 1; other < ordered.length; other += 1)
			{
				const [c2, lo2, hi2] = ordered[other];
				const gap = c2 - c1;
				// Sorted by coordinate, so once the gap is too wide every
				// later run is too.
				if (gap > PAIR_MAX_IN)
				{
					break;
				}
				if (gap < PAIR_MIN_IN)
				{
					continue;
				}
				const overlap = Math.min(hi1, hi2) - Math.max(lo1, lo2);
				if (overlap >= PAIR_OVERLAP_IN)
				{
					const bucket = pyRound(gap * 2) / 2;
					tally[bucket] = (tally[bucket] || 0) + overlap;
				}
			}
		}
	}
	return tally;
}

/**
 * Python sorts tuples element by element, and the pairing above depends on
 * the runs being in coordinate order for its `break` to be sound.
 *
 * @param {Array<number[]>} runs
 */
export function sortRuns(runs)
{
	return runs.slice().sort((one, two) =>
		(one[0] - two[0]) || (one[1] - two[1]) || (one[2] - two[2]));
}

/**
 * How much of a layer's pairing sits on its single commonest gap. High means
 * one repeated thickness, which is construction; low means a scatter of
 * depths, which is cabinetry.
 *
 * Reported but NOT ranked on - see `rank`.
 *
 * @param {Layer} layer
 */
export function concentration(layer)
{
	const values = Object.values(layer.pairs);
	if (!values.length)
	{
		return 0;
	}
	return Math.max(...values) / values.reduce((sum, value) => sum + value, 0);
}

/**
 * Score every layer for how much it looks like the structure of a house.
 *
 * Ink, times how axis-aligned it is, times how much of the plan it reaches -
 * and NOT concentration, which inverts the answer: a run of kitchen cabinets
 * is parallel lines at one repeated depth and scores 1.00, while a real house
 * has several wall thicknesses and scores 0.31.
 *
 * A layer still has to pair up SOMEWHERE at a thickness someone could build,
 * which is what keeps the dimension strings out - they are long, dead straight
 * and reach further than the building, so spread actively rewards them, but
 * they pair at 9in where the walls pair at 3.5.
 *
 * @param {Map<string, Layer>} layers
 * @param {number} pageAreaSqIn
 */
export function rank(layers, pageAreaSqIn)
{
	const scored = [];
	for (const [key, layer] of layers)
	{
		if (layer.segments < MIN_SEGMENTS || !layer.extent)
		{
			continue;
		}
		const [x0, y0, x1, y1] = layer.extent;
		const spread = pageAreaSqIn ? ((x1 - x0) * (y1 - y0)) / pageAreaSqIn : 0;

		let gap = null;
		let best = -Infinity;
		for (const [bucket, weight] of Object.entries(layer.pairs))
		{
			if (weight > best)
			{
				best = weight;
				gap = Number(bucket);
			}
		}
		const buildable = gap !== null && gap >= WALL_GAP_MIN_IN && gap <= WALL_GAP_MAX_IN;
		scored.push({
			key,
			layer,
			spread,
			concentration: concentration(layer),
			structure: buildable ? layer.ink * layer.axisShare * spread : 0,
			modalGap: gap,
		});
	}
	return scored.sort((one, two) => two.structure - one.structure);
}

/**
 * The layer the walls are drawn on, as horizontal and vertical runs.
 *
 * @param {import('./readers/model.js').DrawnPath[]} paths
 * @param {number[]} clip
 * @param {number} scale
 * @param {?string} [override] A layer key, for when the ranking is wrong -
 * which it will be sometimes, and `listLayers` is how a person finds out.
 */
export function structure(paths, clip, scale, override = null)
{
	const found = split(paths, clip, scale);
	if (override)
	{
		const layer = found.get(override);
		if (!layer)
		{
			throw new Error(
				`This drawing has no layer ${override}. Pick one of: `
				+ [...found.keys()].join('  '));
		}
		return {horizontal: layer.horizontal, vertical: layer.vertical, key: override, layers: found};
	}
	const area = (clip[2] - clip[0]) * scale * (clip[3] - clip[1]) * scale;
	const ordered = rank(found, area);
	if (!ordered.length || !ordered[0].structure)
	{
		throw new Error(
			'No layer in this region looks like the structure of a building. '
			+ 'Check the region covers the floor plan, and that the scale is right.');
	}
	return {
		horizontal: ordered[0].layer.horizontal,
		vertical: ordered[0].layer.vertical,
		key: ordered[0].key,
		layers: found,
		ranked: ordered,
	};
}

/**
 * Every layer in the region, best-ranked first, for the layer picker.
 *
 * The ranking is wrong sometimes and is deliberately not tuned to any one
 * drawing, so the interface offers the alternatives rather than hiding them.
 *
 * @param {import('./readers/model.js').DrawnPath[]} paths
 * @param {number[]} clip
 * @param {number} scale
 */
export function listLayers(paths, clip, scale)
{
	const found = split(paths, clip, scale);
	const area = (clip[2] - clip[0]) * scale * (clip[3] - clip[1]) * scale;
	const ranked = rank(found, area);
	const seen = new Set(ranked.map((entry) => entry.key));
	// The ones rank() would not look at are still offered, last and marked,
	// because "too few segments" is a property of the CLIP as much as of the
	// layer - a region drawn tight around one room can put the wall pen below
	// the floor.
	const rest = [...found.entries()]
		.filter(([key]) => !seen.has(key))
		.map(([key, layer]) => ({key, layer, spread: 0, structure: 0, modalGap: null,
			concentration: concentration(layer)}));

	return [...ranked, ...rest].map((entry) => ({
		key: entry.key,
		colour: entry.layer.colour,
		width: entry.layer.width,
		fill: entry.layer.fill,
		segments: entry.layer.segments,
		ink: entry.layer.ink,
		structure: entry.structure,
		modalGap: entry.modalGap,
	}));
}
